### PR TITLE
Three fixes found by running TSan.

### PR DESCRIPTION
**Pull request by mjp41**
These are three issues highlighted by TSan.  This PR also enables TSan and UBSan in CI.

### CODE REVIEW

File: src/mem/localalloc.h [hunk 1]
```diff
@@ -426,7 +426,7 @@ namespace snmalloc
       // make pass through call aligned_alloc with the alignment snmalloc
       // would guarantee.
       void* result = external_alloc::aligned_alloc(
-        natural_alignment(size | (sizeof(void*) - 1)), round_size(size));
+        natural_alignment(size), round_size(size));
```
Comment (nwf) [inline src/mem/localalloc.h]: Did you mean to revert this hunk from 966fe9df74f4b337b11c0ce0e7e08cacfe1716ce?

Comment (mjp41) [reply src/mem/localalloc.h]: Yeah, this was actually incorrect.  I don't know how it passed locally.

File: src/mem/remoteallocator.h [hunk 2]
```diff
@@ -106,9 +106,12 @@ namespace snmalloc
...
       freelist::Object::atomic_store_null(last, key);
 
-      // exchange needs to be a release, so nullptr in next is visible.
+      // Exchange needs to be acq_rel.
+      // *  It needs to be a release, so nullptr in next is visible.
+      // *  Needs to be acquire, so linking into the list does not race with
+      //    the other threads nullptr init of the next field.
       freelist::QueuePtr prev =
-        back.exchange(capptr_rewild(last), std::memory_order_release);
+        back.exchange(capptr_rewild(last), std::memory_order_acq_rel);
```
Comment (mjp41) [inline src/mem/remoteallocator.h:115]: So my investigation of a TSAN error can be found here:
https://gist.github.com/mjp41/3747d988209ca1d0d18a5f09ef9d67c7

I think `acquire` is much safer here.  I think the coherence order on `back` might induce sufficient order to make this actually correct, but moving to `acquire` for now.  We can revisit this choice at a later point if perf is harmed on weaker platforms than x64.

**Review by nwf — APPROVED** [review]
All LGTM.

**Pull request merged**